Add script to gather key Cyber Hygiene metrics

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds a script to gather key Cyber Hygiene metrics from various sources (database and reporter logs, CyHy database).  The script output is a header row and a single line of comma-separated values.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

As our CyHy system grows, we need an easy way to keep tabs on certain key metrics.  This script (`gather_key_cyhy_metrics.sh`) allows us to easily do that.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I have run the script and ensures that it outputs the intended data.  I also confirmed that the usage information is displayed if the incorrect number of parameters are passed to the script.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

dav3r committed:
Initial commit of gather_key_cyhy_metrics.sh

It's a script to gather key Cyber Hygiene metrics from various sources.
Output is a header row and a single line of comma-separated values.

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+  echo "cyhy_reporter_fqdn: The fully qualified domain name of the Cyber Hygiene reporter server (e.g. \"reporter.example.gov\")"
+  echo "cyhy_mongodb_uri: The MongoDB URI for the Cyber Hygiene database (e.g. \"mongodb://localhost:27017/cyhy\" or \"localhost/cyhy\")"
+  echo "cyhy_mongodb_username: The MongoDB username for the Cyber Hygiene database"
+  echo "cyhy_mongodb_password: The MongoDB password for the Cyber Hygiene database"
```
I think a heredoc is a bit less busy. I also put in the parameter expansion snippet to use the calling script name instead.

```suggestion
  cat << END_OF_LINE
Usage:  ${0##*/} cyhy_db_fqdn cyhy_reporter_fqdn cyhy_mongodb_uri cyhy_mongodb_username cyhy_mongodb_password

cyhy_db_fqdn: The fully qualified domain name of the Cyber Hygiene database server (e.g. "database.example.gov")
cyhy_reporter_fqdn: The fully qualified domain name of the Cyber Hygiene reporter server (e.g. "reporter.example.gov")
cyhy_mongodb_uri: The MongoDB URI for the Cyber Hygiene database (e.g. "mongodb://localhost:27017/cyhy" or "localhost/cyhy")
cyhy_mongodb_username: The MongoDB username for the Cyber Hygiene database
cyhy_mongodb_password: The MongoDB password for the Cyber Hygiene database

END_OF_LINE
```

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+# COMMANDER MAIN LOOP DURATION METRICS
+# NOTE: These metrics are pulled from all existing (non-compressed) commander
+# logs, not just the most recent one.
+commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
```
Just a thought to be specific since we're assuming the data is numeric.
```suggestion
commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took \d* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
```

dav3r replied:
The number we are capturing has a decimal and a single digit after the decimal in it, e.g.:
```
Last cycle took 1835.6 seconds.
```
 
Are you good with this instead?
```suggestion
commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took [\d.]* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
```

mcdonnnj replied:
Yep that works for me. Sorry I forgot it's a float and not an int.

dav3r replied:
This change was incorporated as part of https://github.com/cisagov/cyhy-system/pull/125/commits/803f7d0dbef2a5a29d0f57945bb7c6c610a5ddaf.

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+# NOTE: These metrics are pulled from all existing (non-compressed) commander
+# logs, not just the most recent one.
+commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
+commander_duration_median=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | cut --delimiter ' ' --fields 4 | sort --numeric-sort | awk '{ a[i++]=\$1; } END { print a[int(i/2)]; }'")
```
```suggestion
commander_duration_median=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took \d* seconds' /var/log/cyhy/commander.log* | cut --delimiter ' ' --fields 4 | sort --numeric-sort | awk '{ a[i++]=\$1; } END { print a[int(i/2)]; }'")
```

dav3r replied:
This change was incorporated as part of https://github.com/cisagov/cyhy-system/commit/803f7d0dbef2a5a29d0f57945bb7c6c610a5ddaf.

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+# COMMANDER MAIN LOOP DURATION METRICS
+# NOTE: These metrics are pulled from all existing (non-compressed) commander
+# logs, not just the most recent one.
+commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
```
Would it make sense to pull the remote information once and then process it three times locally? This ensures the same values are used for each computation since I assume we are _not_ pausing the commander before running this script.

dav3r replied:
Good thinking - please take a look at https://github.com/cisagov/cyhy-system/commit/803f7d0dbef2a5a29d0f57945bb7c6c610a5ddaf and let me know if that works for you.

mcdonnnj replied:
LGTM ✔

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+# logs, not just the most recent one.
+commander_duration_mean=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | awk '{ total += \$4; count++ } END { print total/count }'")
+commander_duration_median=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | cut --delimiter ' ' --fields 4 | sort --numeric-sort | awk '{ a[i++]=\$1; } END { print a[int(i/2)]; }'")
+commander_duration_max=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took .* seconds' /var/log/cyhy/commander.log* | cut --delimiter ' ' --fields 4 | sort --numeric-sort | tail --lines 1")
```
```suggestion
commander_duration_max=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took \d* seconds' /var/log/cyhy/commander.log* | cut --delimiter ' ' --fields 4 | sort --numeric-sort | tail --lines 1")
```

dav3r replied:
This change was incorporated as part of https://github.com/cisagov/cyhy-system/commit/803f7d0dbef2a5a29d0f57945bb7c6c610a5ddaf.

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,84 @@
+done
+
+# WEEKLY REPORTING METRICS
+weekly_snapshots_duration_minutes=$(ssh "$cyhy_reporter_fqdn" "tail --lines 10 /var/cyhy/reports/output/snapshots_reports_scorecard_automation.log | grep 'Time to generate snapshots' | cut --delimiter ' ' --fields 9")
+weekly_reports_duration_minutes=$(ssh "$cyhy_reporter_fqdn" "tail --lines 10 /var/cyhy/reports/output/snapshots_reports_scorecard_automation.log | grep 'Time to generate reports' | cut --delimiter ' ' --fields 9")
+weekly_total_reporting_duration_minutes=$(ssh "$cyhy_reporter_fqdn" "tail --lines 10 /var/cyhy/reports/output/snapshots_reports_scorecard_automation.log | grep 'Total time' | cut --delimiter ' ' --fields 7")
```
Same question about pulling data once and then processing locally. Obviously with reporting it's unlikely the data will change, but it just feels silly to ssh three times just to retrieve the same base data and then process it on the remote side.

dav3r replied:
Optimized in https://github.com/cisagov/cyhy-system/pull/125/commits/d1192a3d5a57f8ae08c5b69110699ba170a7dc56. ⚡

dav3r committed:
Use a heredoc for script usage text

This is cleaner than a bunch of echo statements.

Also, use parameter expansion for the name of the script in the usage text.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add a helpful URL to a comment

dav3r committed:
Rework commander main loop duration section

* Only one SSH call to fetch log data, instead of three
* Improved regex that specifies a floating point number

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Rework weekly reporting metrics section to only make one SSH call instead of three

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove an unnecessary comment

jsf9k commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,95 @@
+commander_duration_text=$(ssh "$cyhy_db_fqdn" "grep --only-matching --perl-regexp 'Last cycle took [\d.]* seconds' /var/log/cyhy/commander.log*")
+
+commander_duration_mean=$(echo "$commander_duration_text" | awk '{ total += $4; count++ } END { print total/count }')
+# Note: The macOS default "cut" command does not support the long flag names
+# for --delimeter and --fields.
```
Oh for the love of JVPITER!!!

dav3r replied:
That was one reason why I liked doing all of the processing on the remote instance, but here we are.  I still think it's better to do things the way we are doing them now, but agreed that the long flag name thang is annoying.

jsf9k replied:
I agree that the new way is better, but macOS's nonsense is aggravating.

mcdonnnj commented on scripts/gather_key_cyhy_metrics.sh:
```diff
@@ -0,0 +1,95 @@
+set -o errexit
+set -o pipefail
+
+if [ $# -eq 5 ]; then
+  cyhy_db_fqdn=$1
+  cyhy_reporter_fqdn=$2
+  cyhy_mongodb_uri=$3
+  cyhy_mongodb_username=$4
+  cyhy_mongodb_password=$5
+else
```
Just a thought but we could simplify this down to just an `if` `-ne` and eliminate the `else` entirely.

dav3r replied:
Good call- done in https://github.com/cisagov/cyhy-system/pull/125/commits/6dbb2e642502a0040ffdc56478c3266b63ce1b98. 👍

mcdonnnj reviewed (approved):
LGTM ✔ Thanks for putting this together. One last thought for your consideration.

dav3r committed:
Simplify script parameter check conditional logic

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
